File: vimoutliner/scripts/votl_maketags.pl
#!/usr/bin/env perl
# #######################################################################
# votl_maketags.pl: Vim outline tagging system, main program, version 0.3.5
#   Copyright (C) 2001-2003, 2011 by Steve Litt (slitt@troubleshooters.com)
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, see <http://www.gnu.org/licenses/>.
#
# Steve Litt, slitt@troubleshooters.com, http://www.troubleshooters.com
# #######################################################################

# #######################################################################
# #######################################################################
# #######################################################################
# HISTORY
# V0.1.0 Pre-alpha
#     Starting at a "top level" indent-defined Vim outline, this
#     program finds all "tags" defined as headlines starting with
#     _tag_, and containing a subheadline containing the file
#     to which the tag should jump. This program creates a tags
#     file.
#Steve Litt, 5/28/2001
#End of version 0.1.0
#
# V0.1.1 Pre-alpha
#     Bug fixes, including ../ resolution
#
#Steve Litt, 5/28/2001
#End of version 0.1.1
#
#
# V0.1.2 Pre-alpha
#	More bug fixes, and facility to create a new outline
#	file from a tag whose corresponding file doesn't yet
#	exist.  
#Steve Litt, 5/30/2001
#End of version 0.1.2
#
# V0.1.3 Pre-alpha
#	More bug fixes. This was the first version released
#	file from a tag whose corresponding file doesn't yet
#	exist.  
#Steve Litt, 6/01/2001
#End of version 0.1.3
#
# V0.2.0 Pre-alpha
#Steve Litt, 12/03/2002
#	This file unchanged. The overall Vimoutliner version
#	0.2.0 has extensive improvements, including intuitive
#	collapse/expand.
#End of version 0.2.0
#END OF HISTORY
#
#
# V0.1.2 Pre-alpha
#	More bug fixes, and facility to create a new outline
#	file from a tag whose corresponding file doesn't yet
#	exist.  
#Steve Litt, 5/30/2001
#End of version 0.1.2
# V0.3.5 release 20110303
#       Changed vo_tags.tag directory from
#       $HOME/.vimoutliner/
#       to
#       $HOME/.vim/vimoutliner/
#Steve Litt, 3/3/2011
#End of version 0.3.5 release 20110303
#END OF HISTORY
#
# #######################################################################

use strict;
use warnings;
use Path::Tiny;
use autodie qw(:all);
use List::AllUtils qw(uniq);

my $TAGFILENAME = path("$ENV{HOME}/.vim/vimoutliner/vo_tags.tag");
my $TAGFILENAME_fh;

# HashRef containing a map from a filename to its tag names
# { filename => { tagname => filename } }
my $files_to_tags;
# Array of all the files left to process
my @process_queue;

main();

sub main {
	return usage() unless @ARGV;

	push @process_queue, @ARGV; # add all arguments to the process queue

	$TAGFILENAME_fh = $TAGFILENAME->opena();
	# no process each one
	while( @process_queue ) {
		create_and_process( shift @process_queue );
	}
	$TAGFILENAME_fh->close;

	sort_and_dedupe_tagfile();
}

sub usage {
	print "\nUsage is:\n";
	print "otltags topLevelOutlineFileName\n\n";
}

sub expand_filename {
	my ($filename, $base_dir) = @_;
	$filename =~ s|^\$HOME/|~/|; # special case the $HOME environment variable
	$filename = path( $filename ); # path expansion
	$filename = $filename->absolute( $base_dir ) if( $base_dir );
	$filename;
}

sub process_file {
	my ($filename) = @_;
	my $f = path( $filename );
	my $f_contents = $f->slurp;
	my %f_tags = $f_contents =~ /
		^\s*(?<tagname>_tag_\S+).*
		\n # and on the next line
		^\s*(?<filename>.*)
		/mgx;
	\%f_tags;
}

sub create_and_process {
	my ($filename) = @_;
	# we want an absolute path
	$filename = expand_filename( $filename, Path::Tiny->cwd );

	# it has already been processed
	return if exists $files_to_tags->{$filename};

	if( ! -f $filename ) {
		# does not exist: create it
		$filename->parent->mkpath( { verbose => 1 } ); # make path and be verbose about it
		$filename->touch;
		$files_to_tags->{$filename} = {}; # new file -> already processed
	} else {
		my $results = process_file( $filename );
		my $basedir = $filename->parent;
		for my $tag (keys $results) {
			# expand the files for each of the tags
			$results->{$tag} = expand_filename($results->{$tag}, $basedir);
		}

		# and add the files for each of the tags to the @process_queue
		push @process_queue, values %$results;

		append_tags_to_tagfile($results);

		$files_to_tags->{$filename} = $results; # let's store all the tags (useful for debugging)
	}
}

sub sort_and_dedupe_tagfile {
	my @contents = $TAGFILENAME->lines;
	my @uniq_tags = uniq sort { $a cmp $b } @contents;
	$TAGFILENAME->spew( \@uniq_tags );
}

sub append_tags_to_tagfile {
	my ($tags) = @_;
	for my $tag (keys $tags) {
		print $TAGFILENAME_fh "$tag\t$tags->{$tag}\t:1\n"
	}
}

File: vimoutliner/scripts/votl_maketags.py
#!/usr/bin/env python3
#   Based on votl_maketags.pl
#   Copyright (C) 2001-2003, 2011 by Steve Litt (slitt@troubleshooters.com)
#   This script is
#   Copyright (C) 2015 Matěj Cepl <mcepl@cepl.eu>
#
#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, see <http://www.gnu.org/licenses/>.
from __future__ import print_function, unicode_literals

import argparse
import os.path
import re
import errno


TAGFILENAME = os.path.expanduser("~/.vim/vimoutliner/vo_tags.tag")

TAG_RE = re.compile(r'''
^\s*(?P<tagname>_tag_\S+).*
    \n # and on the next line
    ^\s*(?P<filename>.*)
''', re.VERBOSE | re.MULTILINE)


def append_tags_to_tagfile(tags, outfile):
    for tag in tags:
        print("{0}\t{1}\t:1".format(tag, tags[tag]), file=outfile)


def process_file(filename):
    f = os.path.abspath(filename)
    f_contents = open(f, 'r').read()
    f_tags = {}

    for match in TAG_RE.finditer(f_contents):
        f_tags[match.group('tagname')] = \
            match.group('filename')

    return f_tags

"""Open requested file for appending, returning a file  object
Will create any necessary directories and even the target file
if they are not present.
May throw an error if the operations are not possible.
"""
def open_or_make(filename):
    if os.path.exists(filename):
        return open(filename, 'a')
    basedir = os.path.dirname(filename)
    if not os.path.exists(basedir):
        os.makedirs(basedir)
    return open(filename, 'a') # creates empty file
    
def create_and_process(filename, outfile, queue, filestag):
    filename = os.path.abspath(filename)

    if filename in filestag:
        return

    basedir = os.path.dirname(filename)

    if not os.path.exists(filename):
        open_or_make(filename)
        filestag[filename] = {}
    else:
        results = process_file(filename)
        for tag in results:
            target = results[tag]
            target = os.path.expandvars(os.path.expanduser(target))
            if not os.path.isabs(target):
                target = os.path.join(basedir, target)
            results[tag] = os.path.abspath(target)
        queue.extend(results.values())

        append_tags_to_tagfile(results, outfile)

        # let's store all the tags (useful for debugging)
        filestag[filename] = results


def sort_and_dedupe_tagfile():
    sorted_set = sorted(set(open(TAGFILENAME, 'r').readlines()))
    with open(TAGFILENAME, 'w') as tagfile:
        tagfile.writelines(sorted_set)


def main():
    # dict containing a map from a filename to its tag names
    # { filename => { tagname => filename } }
    files_to_tags = {}
    parser = argparse.ArgumentParser()
    parser.add_argument('files', nargs="+",
                        help='otl files to scan for tags')
    args = parser.parse_args()
    process_queue = args.files

    tagfile = open_or_make(TAGFILENAME)

    for otl_file in process_queue:
        create_and_process(otl_file, tagfile, process_queue,
                           files_to_tags)

    tagfile.close()
    sort_and_dedupe_tagfile()


if __name__ == '__main__':
    main()
